Fix reentrant GCS::update()

lucasdemarchi opened the pull request:
See commit message with detailed backtrace.

This also changes the Scheduler::delay() call in AP_HAL_Linux so it only allows calls from the main thread.

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -38,6 +38,7 @@ uint32_t GCS_MAVLINK::reserve_param_space_start_ms;
 
 AP_HAL::Util::perf_counter_t GCS_MAVLINK::_perf_packet;
 AP_HAL::Util::perf_counter_t GCS_MAVLINK::_perf_update;
+bool GCS_MAVLINK::_in_update = false;
```
This really has to be a semaphore.  Currently there's a TOCTTOU issue here.

lucasdemarchi replied:
Why? There's no multithread involved in this function. If it's called by another thread we already lost.

peterbarker commented on libraries/GCS_MAVLink/GCS.h:
```diff
@@ -323,6 +324,8 @@ class GCS_MAVLINK
     // this allows us to detect the user wanting the CLI to start
     uint8_t        crlf_count;
 
+    static bool _in_update;
```
Making this static will mean backends which might have been able to get work done will not be able to.

lucasdemarchi replied:
Making it static was me being lazy... It's not worse than it current it is due to the other variables used on this function being static

lucasdemarchi commented:
It's not. It's just the function name that is misleading/wrong. AP_HAL_PX4 checks it's not the main thread. I'll add a commit to fix the function name:

```
bool PX4Scheduler::in_timerprocess()
{
    return getpid() != _main_task_pid;
}
```

peterbarker commented:
@lucasdemarchi Gah!  I even remember noticing that before!

peterbarker commented:
On Tue, 1 Aug 2017, Lucas De Marchi wrote:

> Why? There's no multithread involved in this function. If it's called by
> another thread we already lost.

You're correct, only the main thread should be touching this variable.

... but why are we actually tring to stop update being reentered again?

lucasdemarchi commented:
> ... but why are we actually tring to stop update being reentered again?

See the backtrace on the commit message. In this particular case it was:
1. GCS asks to test motors
1. ArduCopter handles the mavlink message (thus inside the GCS::update() method)
1. During motor initialization it calls hal.scheduler->delay()
1. The delay function will end up "processing mavlink packets" if delay is greater than 5msec to avoid it being locked out
1. That brings us back into GCS::update()

All of that happens on the same (main) thread.

peterbarker commented:
@lucasdemarchi Yes, I can see that.  Is that a problem, 'though?  You may get useful work done during that delay.

OXINARF commented:
@peterbarker You need to sleep more 😛 If you enter update twice there is a chance you will be reading a packet half in the first call and half in the second - that's not going to end up well!

@lucasdemarchi Can you split the first commit? It is fixing the reentrance but also refactoring waypoint receiving to be a separate method - I like it, but should be in a separate commit.

lucasdemarchi commented:
@OXINARF @peterbarker just did the requested changes. Hold on on merging since it's only build tested.

OXINARF commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -61,11 +59,13 @@ GCS_MAVLINK::init(AP_HAL::UARTDriver *port, mavlink_channel_t mav_chan)
     }
-    if (!_perf_update) {
-        _perf_update = hal.util->perf_alloc(AP_HAL::Util::PC_ELAPSED, "GCS_Update");
+    if (snprintf(perf_name, sizeof(perf_name), "GCS_Packet_%u", chan) == 0) {
```
GCS_Packet_%u -> GCS_Update_%u

OXINARF commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -890,14 +896,13 @@ GCS_MAVLINK::update(run_cli_fn run_cli, uint32_t max_time_us)
 
+    if (!waypoint_receiving) {
+        goto end;
+    }
```
Nitpicking: don't change the position of this, it will go to the previous position in a later commit.

lucasdemarchi commented:
@OXINARF thanks. Fixed your comments, tested and pushed again.  During test I also realized the check in snprintf was wrong and removed it:  first because it won't overflow, second because it's better a truncated name than failing the check and using the wrong (0) perf counter.

tridge commented:
I have some concerns about the reentrant fix. We rely on mavlink keeping working during some long-running tasks.
I would be OK with preventing more than one COMMAND_LONG or COMMAND_INT running at a time, which I believe would fix the issue, but stopping all mavlink processing is not a good move

lucasdemarchi commented:
@tridge, Besides the problem with the perf counters that's being fixed here, there's another not so obvious that @oxinarf mentioned: one iteration could get a partial part of the message and screw the other one.

Long running tasks should not block the main loop. This is only guaranteeing a bad part of the code doesn't cause even more harm.

tridge commented:
> one iteration could get a partial part of the message and screw the other one.

how can that happen? It isn't multi-threading - the callback always happens when the caller is quiescent, and has fully parsed a message.

OXINARF commented:
@lucasdemarchi, @tridge is right. Delay will only be called in the main thread and it should also only be called after _packetReceived_ is called. At that point the message has been put into the mavlink message struct and out of the buffer - re-entrance shouldn't be a problem here.
One problem about putting this fix in is that it can break code that relies on MAVLink communication while not being async - meaning it would rely on the current re-entrant behavior.

I can't see any other immediate problem now, but I'll look at the code with more attention tomorrow.

@lucasdemarchi @tridge I think that the Scheduler commits are good to go in.

lucasdemarchi commented:
I was not worried with multithread, I know it's on the main thread.  I had read the update function was trying to parse the message using a buffer on stack, but it actually parses using a shared buffer. So the problem of partially read messages should not happen indeed.

I still think this is a can of worms:
- How to fix the perf counters? Moving the "parsed_packet" to the class and don't count if we are reentrant would be a possibility... ugly I must say
- The "max time" arg in this function completely loses purposes
- It makes it more difficult to develop the other side: due to reordering messages (messageA; messageB; answerB; answerA)
- Let's remind this is all going on on the main thread, so whatever code is relying is a candidate to change the algorithm to be async rather than sync

lucasdemarchi commented:
I pushed the fixes to Scheduler::delay() on Linux and removal of in_timer_process().  The rest remains here to be discussed.

OXINARF commented:
@lucasdemarchi I agree with all your points, but we can't just break an unknown number of features by closing the hole. I think we need to first identify all cases where closing the hole would pose an issue and then analyze what can be done.

lucasdemarchi commented:
@OXINARF I would be really surprised if we have any.    The one mentioned here (motor test) works perfectly fine.

Btw, it's not just these perf counters that are broken. Any perf counter triggered by the re-entrance is broken as well :(

lucasdemarchi commented:
In order for me to check what could potentially be broken I can check all the calls to `Scheduler::delay()` that sleeps for more than 5msec and are called on the main thread.   We have lots of calls to delay, but not many with `>= 5msec`. Most of them are easy to track down coming from the init sequence, so it shouldn't be re-entrant neither.

OXINARF commented:
@lucasdemarchi If you can check that, it would be great.

tridge commented:
the delays I'm particularly concerned about are the ones that may take an extended time. For example, the delay in AP_InertialSensor::_init_gyro(), which can run on for a long time if the vehicle is not still. That delay is deliberately 5ms in order to trigger this re-entrant logic so that parameter download continues while gyro is calibrating. Breaking that would affect a lot of users quite badly.
We could make the _init_gyro() code fully async, and that would help a lot. 
If we can find and fix the places that do these delays (eg. gyro init) so they are async then I'd be happy to remove that whole re-entrant logic.

lucasdemarchi commented:
@tridge my point is that in the init path there's no re-entrance. It only happens if you are handling a mavlink packet and you call delay with more than 5msec. Those can be made async and I don't think we have many.

tridge commented:
> my point is that in the init path there's no re-entrance. 

I don't quite understand then. The gyro init happens on the init path (though it can happen after init as well)

> Those can be made async and I don't think we have many.

yes I agree, but we can't make this change to disable the re-entrancy until we have removed all the places this can happen. The most obvious one to start with is the gyro init, as that affects every board on every flight.

lucasdemarchi commented:
See ArduPilot/ardupilot/pull/6704/commits/47372c5e08eee9ff960f492806ae7adefc99b39d

I don't know where the issue with baro initialization comes from... It is not what this PR is fixing. This is only fixing re-entrancy when handling mavlink packets.

lucasdemarchi commented:
I applied the first commit  from this PR that makes the perf counters non-static so it at least works with different channels.

lucasdemarchi commented:
I'm reviewing all places that call delay(x), with x >= 5. We can ignore all cases in which the call chain isn't one triggered by receiving a mavlink packet.

So far:
- handling MAV_CMD_PREFLIGHT_REBOOT_SHUTDOWN in Copter and Sub
- Copter::mavlink_compassmot (call to delay() and to init_rc_out())
- Copter::mavlink_motor_test_start() that calls init_rc_out())
- Copter::esc_calibration_auto() that's called by init_rc_out())
- PX4Scheduler::reboot()
- GCS_MAVLINK::handle_preflight_reboot()
- GCS_MAVLINK::handle_serial_control()


Others that are easy to mistake by a false positive (I did):
- init_barometer() -> AP_Baro::calibrate() -> delay(10) / delay(100).  init_barometer() can be called from a call chain handling a mavlink packet, but never doing a full calibration (so it doesn't call AP_Baro::calibrate().
- AP_BoardConfig::sensor_config_error(): it calls delay() to let mavlink messages keep going and stay stuck in a loop... but this is only triggered on initialization, not from mavlink

lucasdemarchi commented:
So, I actually don't see any critical code path. esc calibration and motor test all continue to work with the exception of not sending/receiving mavlink messages for a little bit more.  motortest is something to investigate as I am not familiar with what it is doing in that place.

lucasdemarchi commented:
Rebase done.  I left it with only one commit as one commit had already been applied and the split of _waypoint_receice() didn't make sense anymore since we are using tnow to send TIMESYNC.

I also added a comment in the code explaining what this is fixing.

WickedShell commented:
@lucasdemarchi  I assume you weren't trying to close this? (Reopened on that assumption, feel free to close it if you really intended to push then close a PR :D )

lucasdemarchi commented:
@WickedShell 

Somehow the grey button was more attractive to my eyes than the green one o.O

Thanks.

peterbarker commented:
I think it will take us quite some time to get rid of all of the time-consuming synchronous things we do from `::update()`.  Perhaps in the meantime we could guard the manipulation of the perf counters with `_in_update` rather than denying re-entrancy.

lucasdemarchi commented:
@peterbarker  can you point to a single example? See my comment and my analysis above: there shouldn't be any case like you say today.

lucasdemarchi commented:
To be clear how to check: grep for all the calls to delay() that are above the threshold. We don't really have many. For each location, check if we can be there due to be processing a mavlink packet. The conclusion is in https://github.com/ArduPilot/ardupilot/pull/6704#issuecomment-328997143 and  we should be safe already.

Removing the synchronous things from the main thread is good but doesn't have much to do with this patch.

tridge commented:
@lucasdemarchi the approach I'd like to take with this is to move all mavlink sending to its own thread, which will greatly simplify the logic, completely remove the need for the out_of_time stuff, remove the need for the whole delay callback mechanism (as the thread will be running all the time), plus it will allow us to run the main loop on copters a lot faster (the main constraint on loop speed at the moment is the mavlink code).
I have started on some new approaches to semaphores to make putting mavlink in its own thread cleaner and easier

lucasdemarchi commented:
I'll close this one then.
